File: app/src/main/java/com/project/scratchstudio/kith_andoid/network/apiService/UserApi.java
package com.project.scratchstudio.kith_andoid.network.apiService;

import com.project.scratchstudio.kith_andoid.network.model.BaseResponse;
import com.project.scratchstudio.kith_andoid.network.model.referral.ReferralResponse;
import com.project.scratchstudio.kith_andoid.network.model.user.User;
import com.project.scratchstudio.kith_andoid.network.model.user.UserListResponse;
import com.project.scratchstudio.kith_andoid.network.model.user.UserResponse;

import io.reactivex.Single;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;

public interface UserApi {

    @GET("api/users/{id}")
    Single<UserResponse> getUser(@Path("id") int id);

    @FormUrlEncoded
    @POST("api/users/referral")
    Single<ReferralResponse> getReferralCode(@Field("user_id") int id);

    @FormUrlEncoded
    @POST("api/users/search")
    Single<UserListResponse> searchUsers(@Field("user_id") int id, @Field("search") String search, @Field("page") String page,
                                         @Field("size") String size);

    @FormUrlEncoded
    @POST("api/users/referral_users")
    Single<User[]> getInvitedUsers(@Field("user_id") int id, @Field("page") String page,
                                   @Field("size") String size);

    @FormUrlEncoded
    @POST("api/users/edit")
    Single<BaseResponse> editUser(@Field("user_id") int id, @Field("user_firstname") String firstName, @Field("user_lastname") String lastName,
                                  @Field("user_middlename") String middleName, @Field("user_phone") String phone, @Field("user_email") String email,
                                  @Field("user_position") String position, @Field("user_description") String description,
                                  @Field("user_photo") String photo);
}

File: app/src/main/java/com/project/scratchstudio/kith_andoid/network/apiService/UserApi.kt
package com.project.scratchstudio.kith_andoid.network.apiService

import com.project.scratchstudio.kith_andoid.network.model.BaseResponse
import com.project.scratchstudio.kith_andoid.network.model.referral.ReferralResponse
import com.project.scratchstudio.kith_andoid.network.model.user.User
import com.project.scratchstudio.kith_andoid.network.model.user.UserListResponse
import io.reactivex.Single
import okhttp3.MultipartBody
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.http.GET
import retrofit2.http.Multipart
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Part
import retrofit2.http.Path
import retrofit2.http.Query

interface UserApi {

    @GET("api/users/{id}")
    fun getUser(@Path("id") id: Int): Single<User>

    @FormUrlEncoded
    @POST("api/users/referral")
    fun getReferralCode(@Field("user_id") id: Int): Single<ReferralResponse>

    @FormUrlEncoded
    @POST("api/users/referral_count")
    fun getReferralCount(@Field("user_id") id: Int, @Field("only_mine") isMine: Int?): Single<ReferralResponse>

    @GET("api/users")
    fun searchUsers(@Query("search") search: String, @Query("page") page: String,
                    @Query("size") size: String): Single<UserListResponse>

    @FormUrlEncoded
    @POST("api/users/referral_users")
    fun getInvitedUsers(@Field("user_id") id: Int, @Field("page") page: String,
                        @Field("size") size: String): Single<UserListResponse>

    @FormUrlEncoded
    @PUT("api/users/{id}")
    fun editUser(@Path("id") userId: Int, @Field("firstname") firstName: String, @Field("lastname") lastName: String,
                 @Field("middlename") middleName: String, @Field("phone") phone: String, @Field("email") email: String,
                 @Field("position") position: String, @Field("description") description: String): Single<BaseResponse>

    @Multipart
    @PUT("api/users/{id}")
    fun editUserPhoto(@Path("id") userId: Int, @Part surveyImage: MultipartBody.Part?): Single<BaseResponse>

    @FormUrlEncoded
    @PUT("api/users/{id}")
    fun changeUserPassword(@Path("id") userId: Int, @Field("password") password: String): Single<BaseResponse>

}
